### PR TITLE
SUMO-75114 upgrade the elasticsearch to scala 2.12 and fix compilation errors

### PR DESCRIPTION
**Pull request by yuting-liu**
###### Description
Upgrade to scala 2.12

@xicmiah

### CODE REVIEW

File: elasticsearch-akka/pom.xml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
   <modelVersion>4.0.0</modelVersion>
 
-  <artifactId>elasticsearch-akka</artifactId>
+  <artifactId>elasticsearch-akka_${scala.version.major}</artifactId>
```
Comment (cddude229) [inline elasticsearch-akka/pom.xml:4]: I think `readme.md` needs changes here to support this new naming.

File: elasticsearch-aws/pom.xml [hunk 2]
```diff
@@ -48,11 +48,10 @@
 
     <!-- Non-Sumo Dependencies. -->
     <dependency>
-      <groupId>io.spray</groupId>
-      <artifactId>spray-can_${scala.version.major}</artifactId>
-      <version>1.3.3</version>
+      <groupId>com.typesafe.akka</groupId>
```
Comment (cddude229) [inline elasticsearch-aws/pom.xml:51]: I'd consider changing pulled in binaries to be a breaking change... however, I don't know best practices here.  We should research a bit to see if that suggests bumping the version of our client

File: pom.xml [hunk 3]
```diff
@@ -25,16 +25,27 @@
     </scm>
 
     <properties>
-        <scala.version.major>2.11</scala.version.major>
-        <akka.version>2.3.11</akka.version>
+        <scala.version.major>2.12</scala.version.major>
```
Comment (cddude229) [inline pom.xml:28]: We can't commit this without a major version bump, IMO

Comment (rcoh) [reply pom.xml:28]: If we care about supporting 2.11, we need to do a cross-version compile where we produce 1 binary per scala version. I've done it in SBT, but I'm not sure how it's done in maven. 

My suggestion would just be to do a major version bump.

File: elasticsearch-akka/pom.xml [hunk 4]
```diff
@@ -52,8 +52,8 @@
 
     <dependency>
       <groupId>com.typesafe.akka</groupId>
-      <artifactId>akka-stream-experimental_${scala.version.major}</artifactId>
```
Comment (CCheSumo) [inline elasticsearch-akka/pom.xml:55]: Glad to see this change finally.

**Comment by CCheSumo** [comment]
Looks good in general. +1 on bumping up the major version. Here are the best practices http://kylelieber.com/2012/06/maven-versioning-strategy/

**Comment by xicmiah** [comment]
I assumed version gets bumped during actual release outside of this PR, together with removing `-SNAPSHOT`.
Ok, will change.

**Comment by CCheSumo** [comment]
@xicmiah I am okay with bumping up the major version in an other PR. Although, I would prefer to do it here so that there is no chance someone releases a version in between. 

No need to remove `-SNAPSHOT`, it will be taken care of during the release process.

**Comment by CCheSumo** [comment]
I can help with the release and following PR to set the right versions after release.

**Comment by szymonm** [comment]
What's the status of this PR? ARe we going to work on it anymore?

I'm interested in bumping akka-streams version

**Comment by coertg** [comment]
Hey folks. Any idea when scala 2.12 support will be available?

**Comment by cddude229** [comment]
@coert-work I'm starting to pick up the 2.12 upgrade work again, so this is on my list - but I don't have an ETA yet.  I welcome any help though if you'd like to contribute.

**Comment by coertg** [comment]
> @coert-work I'm starting to pick up the 2.12 upgrade work again, so this is on my list - but I don't have an ETA yet. I welcome any help though if you'd like to contribute.

My todo list is pretty long as well, but I will def try to get to this. thanks

**Comment by supritha-amudhu-zendesk** [comment]
Edited: @cddude229 I'm trying to follow this thread to see if SumoLogic/elasticsearch-client supports Scala 2.12. Is there an update on this?

Thanks @coertg!

**Comment by coertg** [comment]
> I'm trying to follow this thread to see if SumoLogic/elasticsearch-client supports Scala 2.12. Is there an update on this?

I haven't been able to get to this. It may be worth asking @cddude229

**Comment by cddude229** [comment]
@supritha-amudhu-zendesk @coertg This is still on my roadmap, but I've not been able to get time to focus on it yet, despite wanting to.

My current plan:
1. Get gradle dual publishing working with another open source library (just because I started there)
2. Port this project to use gradle + that dual publishing
3. Upgrade to 2.12 (aka revive this PR itself.)

At that point, we will support both 2.11 and 2.12 of this repo for the time being.  (We will eventually EOL 2.11 support, but we've not yet internally had a discussion on when that'll be.)

In the meantime, if you have many of the fixes need to be 2.11 / 2.12 compatible (besides libraries), so if someone wants to pick those up, that'd be a huge help!

Edit: Also, sorry about the delay - I was on vacation.

**Comment by fdaca** [comment]
Finished Gradle migration, that supports maintenance of multiple Scala versions: https://github.com/SumoLogic/elasticsearch-client/pull/201

**Comment by cddude229** [comment]
Closing since other PR is merged

**Pull request closed without merging**